parserator_web/views.py

```diff
@@ -13,12 +13,42 @@ class Home(TemplateView):
 class AddressParse(APIView):
     renderer_classes = [JSONRenderer]
 
+    # parse requested address string and return components to frontend 
+    #       input: request {input_string; String} from frontend
+    #       output: input_string [String], 
+    #               address_components [Dictionary {address_part: tag}], 
+    #               address_type [String]
     def get(self, request):
-        # TODO: Flesh out this method to parse an address string using the
-        # parse() method and return the parsed components to the frontend.
-        return Response({})
+        # get user's input_string from request data
+        input_string = request.query_params['input_string']
 
+        # 'initalize' the response variables
+        address_components = {}
+        address_type = ""
+
+        # pass request address to parse method
+        try:
+            address_components, address_type = self.parse(input_string)
+        except Exception as e:
```

Comment:
If you look at `e` here there are more details on the error. For example, by inserting `print(str(e))` in this except block and submitting `123 main st chicago il 123 main` to the form I can see:

```
parserator  | ERROR: Unable to tag this string because more than one area of the string has the same label
parserator  |
parserator  | ORIGINAL STRING:  123 main st chicago il 123 main
parserator  | PARSED TOKENS:    [('123', 'AddressNumber'), ('main', 'StreetName'), ('st', 'StreetNamePostType'), ('chicago', 'PlaceName'), ('il', 'StateName'), ('123', 'AddressNumber'), ('main', 'StreetName')]
parserator  | UNCERTAIN LABEL:  AddressNumber
parserator  |
parserator  | When this error is raised, it's likely that either (1) the string is not a valid person/corporation name or (2) some tokens were labeled incorrectly
parserator  |
parserator  | To report an error in labeling a valid name, open an issue at https://github.com/datamade/usaddress/issues/new - it'll help us continue to improve probablepeople!
parserator  |
parserator  | For more information, see the documentation at https://usaddress.readthedocs.io/
```

Could you capture this error message and display it on the frontend in a way that's useful to a user?

Reply:
Yeah, I was struggling with figuring out what message to pass on to the user and was not sure how much information was too much. In this commit [here](https://github.com/noah-marcus/code-challenge/commit/e0aacf7d9177cbb75602b3a2210b05aafdd3358c), I am now capturing the entire message and displaying it on the frontend. I think the benefit to this approach is that the user will always receive a detailed and specific message from the backend.

Also, I'm not sure how to capture this new message for the unit test check? The commit above is a work in progress because the error handling is failing. @beamalsky, any advice?

Reply:
There's some room for personal style here, and you don't know who your users are in this theoretical app so what you included as an error message is definitely reasonable. 

One way I've handled returning the full error code without overwhelming users is to have a human-friendly error on top, and then the error stashed in under a ["See more" collapsed section](https://getbootstrap.com/docs/4.0/components/collapse/).

Instead of returning a `ParseError` for a bad request, you could return a `Response` with a 400 error code. Something like this:

```python
return Response({
	'input_string': addr, 
	'address_components': '', 
	'address_type': '', 
	'error' : str(e)}, 
	status=400
)
```

That way, you'll have access to the full error text from `usaddress` under `error` and a `Response` to test the attributes of in your test in a way that mirrors successful submission.

Reply:
That makes a lot of sense, thanks for clearing that up! I updated my code in this [commit](https://github.com/noah-marcus/code-challenge/commit/d6af2c9f7b2fdd2956f1bb48d46fe3bd21e291ad) and the test is working again!